File: TRAQS Scheduling/TRAQS Scheduling/Services/ThemeSettings.swift
import SwiftUI
#if canImport(UIKit)
import UIKit
#endif

// MARK: - Background Presets

struct BgPreset: Identifiable {
    let id: Int
    let name: String
    let bg: String
    let surface: String
    let card: String
    let border: String
    let text: String
    let muted: String
    /// Unfilled part of a progress ring or bar. Per-preset rather than one fixed
    /// grey: on glass, a mid-grey track reads as a dirty smudge on a light
    /// surface and disappears entirely on a dark one.
    ///
    /// Each preset pushes AWAY from mid-grey, toward its own extreme — near-white
    /// on White, near-black on Charcoal. On Charcoal that puts the track *below*
    /// the surface value, so it reads as a recessed groove rather than a raised
    /// grey band.
    let track: String
    let isLight: Bool
}

// MARK: - ThemeSettings

@Observable
final class ThemeSettings {

    // Accent presets (sky is the canonical TRAQS interactive color)
    static let accentPresets: [String] = [
        "#3B82F6", // Sky (default) — TRAQS Light
        "#7c3aed", // Purple
        "#10b981", // Green
        "#f59e0b", // Amber
        "#f43f5e", // Red
        "#FF1FB4", // Magenta
        "#06b6d4", // Cyan
        "#8b5cf6", // Violet
    ]

    // Background presets — neutrals only. Accent is what users customize
    // for color; the background stays out of the way as a neutral canvas.
    static let bgPresets: [BgPreset] = [
        BgPreset(id: 100, name: "White",
                 bg: "#F4F6FA", surface: "#FFFFFF", card: "#FFFFFF", border: "#E6E8EE",
                 text: "#0B0B0C", muted: "#6E6E73", track: "#F7F9FD", isLight: true),
        BgPreset(id: 11,  name: "Charcoal",
                 bg: "#1F1F1F", surface: "#2A2A2A", card: "#333333", border: "#3F3F3F",
                 text: "#E8E8E8", muted: "#9CA3AF", track: "#202023", isLight: false),
    ]

    static let defaultBgPresetId: Int = 100
    static let defaultAccent: String = "#3B82F6"
    /// On by default — the liquid wash IS the intended look of the app; the
    /// static canvas is the opt-out.
    static let defaultLiquidBackground: Bool = true
    /// Likewise for glass: on is the intended look, solid is the opt-out.
    static let defaultFrostedGlass: Bool = true

    var accent: String = ThemeSettings.defaultAccent
    var bgPresetId: Int = ThemeSettings.defaultBgPresetId
    /// Whether pages render the drifting liquid wash (`PageBackground`) instead
    /// of the static ambient canvas. Unlike accent/preset this feeds no T.*
    /// token — views read it directly — so it has no `applyToT` counterpart.
    var liquidBackground: Bool = ThemeSettings.defaultLiquidBackground
    /// Whether the app's own SURFACES are frosted glass or flat 2D.
    ///
    /// Off flattens what TRAQS draws: job cards, page boxes, message bubbles,
    /// list rows, the wells inside popups, the sync pill — all to opaque
    /// `T.surface` — and collapses the specular rim on them to a flat hairline.
    /// (`glassFill`, `GlassSurface`, `specularRim`.)
    ///
    /// It does NOT reach three things, each for its own reason:
    ///
    ///  • CONTROLS. Header buttons, menu buttons, the keypad keys and every
    ///    glass CTA are Apple's material, not ours. A flat app with native glass
    ///    controls is a coherent look; one whose buttons went flat too just
    ///    looks unfinished. (`GlassControl`, `GlassCircleButton`, `GlassCTA`.)
    ///
    ///  • THE NAV BAR. It floats over every page, and the page showing through
    ///    it is what says so. Opaque, it reads as a chunk cut out of the screen.
    ///
    ///  • THE PROMPTING POPUPS — the PIN pad, the break/lunch banner, the
    ///    end-job photo prompt, the start-job and time-off confirms. There the
    ///    glass is what signals the thing is floating OVER the page rather than
    ///    being part of it, so it's carrying meaning rather than decoration.
    ///    Those pass `always: true` to the rim and never call `glassFill()`.
    ///    See `GlassPanel`.
    ///
    /// It governed page CONTENT only at first, which left the rim on everything —
    /// a lit bevel being the most obviously glassy thing left once the blur is
    /// gone — and then briefly reached everything, which took the native glass
    /// off the controls too. This is the line that landed.
    ///
    /// Mirrored into T.glassEnabled because the glass helpers include a Shape
    /// extension, which has no view context and so can't read @Environment.
    var frostedGlass: Bool = ThemeSettings.defaultFrostedGlass
    var version: Int = 0

    // Last *saved* theme, captured when the customizer opens (`beginPreview`).
    // Live edits change `accent`/`bgPresetId`/`liquidBackground` for an immediate
    // preview without persisting; Save commits them, backing out reverts to
    // these snapshots.
    private var savedAccent: String = ThemeSettings.defaultAccent
    private var savedBgPresetId: Int = ThemeSettings.defaultBgPresetId
    private var savedLiquidBackground: Bool = ThemeSettings.defaultLiquidBackground
    private var savedFrostedGlass: Bool = ThemeSettings.defaultFrostedGlass

    var currentBgPreset: BgPreset {
        ThemeSettings.bgPresets.first(where: { $0.id == bgPresetId }) ?? ThemeSettings.bgPresets[0]
    }

    var isLightTheme: Bool { currentBgPreset.isLight }

    init() {
        accent = UserDefaults.standard.string(forKey: "themeAccent") ?? ThemeSettings.defaultAccent
        // Any preset id that isn't one of the four current neutrals falls
        // back to White. Covers existing users who were on the older
        // tinted presets (Midnight, Navy, Slate, Forest, Frost, Pearl,
        // Silver, Linen) before we trimmed the list.
        if let savedId = UserDefaults.standard.object(forKey: "themeBgPreset") as? Int,
           ThemeSettings.bgPresets.contains(where: { $0.id == savedId }) {
            bgPresetId = savedId
        } else {
            bgPresetId = ThemeSettings.defaultBgPresetId
        }
        // `object(forKey:) as? Bool`, NOT `bool(forKey:)` — the latter returns
        // false for a key that was never written, which would silently ship the
        // liquid wash OFF for every existing user.
        liquidBackground = (UserDefaults.standard.object(forKey: "themeLiquidBackground") as? Bool)
            ?? ThemeSettings.defaultLiquidBackground
        frostedGlass = (UserDefaults.standard.object(forKey: "themeFrostedGlass") as? Bool)
            ?? ThemeSettings.defaultFrostedGlass
        savedAccent = accent
        savedBgPresetId = bgPresetId
        savedLiquidBackground = liquidBackground
        savedFrostedGlass = frostedGlass
        applyToT()
    }

    /// Live preview only: update the in-memory accent + T tokens so the
    /// customizer reflects the change immediately. Does NOT persist — call
    /// `commitChanges()` (Save) to keep it, or `cancelPreview()` to revert.
    func setAccent(_ hex: String) {
        accent = hex
        applyAccentToT()
    }

    /// Live preview only (see `setAccent`). Persists on `commitChanges()`.
    func setBgPreset(_ id: Int) {
        bgPresetId = id
        applyBgToT(currentBgPreset)
    }

    /// Live preview only (see `setAccent`). Persists on `commitChanges()`. No
    /// `applyToT` call — this flag isn't part of the token table.
    func setLiquidBackground(_ on: Bool) {
        liquidBackground = on
    }

    /// Live preview only (see `setAccent`). Persists on `commitChanges()`.
    func setFrostedGlass(_ on: Bool) {
        frostedGlass = on
        applyGlassToT()
    }

    func reset() {
        setAccent(ThemeSettings.defaultAccent)
        setBgPreset(ThemeSettings.defaultBgPresetId)
        setLiquidBackground(ThemeSettings.defaultLiquidBackground)
        setFrostedGlass(ThemeSettings.defaultFrostedGlass)
        commitChanges()
    }

    /// Snapshot the saved theme before a live-preview session so an
    /// un-saved exit can be reverted.
    func beginPreview() {
        savedAccent = accent
        savedBgPresetId = bgPresetId
        savedLiquidBackground = liquidBackground
        savedFrostedGlass = frostedGlass
    }

    /// Revert a live preview back to the last saved theme (customizer closed
    /// without Save).
    func cancelPreview() {
        accent = savedAccent
        bgPresetId = savedBgPresetId
        liquidBackground = savedLiquidBackground
        frostedGlass = savedFrostedGlass
        applyToT()
    }

    /// Save: persist the previewed accent + background, then bump `version`
    /// so the whole app re-renders with the new T.* values.
    func commitChanges() {
        UserDefaults.standard.set(accent, forKey: "themeAccent")
        UserDefaults.standard.set(bgPresetId, forKey: "themeBgPreset")
        UserDefaults.standard.set(liquidBackground, forKey: "themeLiquidBackground")
        UserDefaults.standard.set(frostedGlass, forKey: "themeFrostedGlass")
        savedAccent = accent
        savedBgPresetId = bgPresetId
        savedLiquidBackground = liquidBackground
        savedFrostedGlass = frostedGlass
        version += 1
    }

    private func applyToT() {
        applyAccentToT()
        applyBgToT(currentBgPreset)
        applyGlassToT()
    }

    private func applyGlassToT() {
        T.glassEnabled = frostedGlass
    }

    /// Set `T.accent` AND the derived signature-gradient stops + glow tints so the
    /// whole gradient system stays coherent with whatever accent is chosen.
    /// EVERY accent — including the default — yields a SAME-HUE two-stop gradient
    /// (the chosen color → a deeper shade of it), so buttons/CTAs are always a
    /// gradient of the exact color chosen, never an off-hue end that reads as a
    /// completely different color.
    private func applyAccentToT() {
        T.accent = accent
        T.accentGradientStart = accent
        T.accentGradientEnd   = ThemeSettings.derivedEnd(from: accent)
        T.glowBlob     = T.accentGradientEnd
        T.ctaGlowColor = T.accentGradientStart
    }

    private func applyBgToT(_ p: BgPreset) {
        T.bg = p.bg; T.surface = p.surface; T.card = p.card; T.border = p.border
        T.text = p.text; T.muted = p.muted
        T.progressTrack = p.track
        applyRimToT(isLight: p.isLight)
    }

    /// The glass edge, tuned per preset family — see the `T.rim*` block.
    ///
    /// Same shape both ways: white glare across the top lip, a darker band down
    /// the sides for contrast, the bottom lip lit again. What changes is how
    /// hard each has to work. On White the lips run near-full white and the side
    /// band is a definite grey, because a subtle edge on a near-white card is no
    /// edge at all. On Charcoal the lips can ease off and the band drops BELOW
    /// the surface colour, so the sides read as a recessed groove.
    private func applyRimToT(isLight: Bool) {
        if isLight {
            T.rimTop   = 0.95
            T.rimBot   = 0.80
            T.rimSide  = "#A6ADB9"
            T.rimLip   = 0.20
            T.rimWidth = 1.4
        } else {
            T.rimTop   = 0.70
            T.rimBot   = 0.50
            T.rimSide  = "#151515"
            T.rimLip   = 0.18
            T.rimWidth = 1.2
        }
    }

    /// Derive a gradient end-stop from the chosen accent: KEEP the hue (no
    /// rotation — that produced an off-hue end that read as a different color),
    /// deepen it into a richer shade (a little more saturation, ~22% less
    /// brightness) so a single color still yields a coherent same-hue two-stop
    /// gradient. iOS-only (UIColor HSB); returns the input unchanged on failure.
    static func derivedEnd(from hex: String) -> String {
        #if canImport(UIKit)
        let ui = UIColor(Color(hex: hex))
        var h: CGFloat = 0, s: CGFloat = 0, b: CGFloat = 0, a: CGFloat = 0
        guard ui.getHue(&h, saturation: &s, brightness: &b, alpha: &a) else { return hex }
        s = min(1.0, s + 0.10)
        b = max(0.0, b - 0.22)
        return Color(UIColor(hue: h, saturation: s, brightness: b, alpha: 1)).toHex() ?? hex
        #else
        return hex
        #endif
    }
}

File: traqs-android/app/src/main/java/com/matrixsystems/traqs/services/ThemeSettings.kt
package com.matrixsystems.traqs.services

import android.content.Context
import androidx.lifecycle.ViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow

// MARK: - Background Presets
// Mirrors iOS ThemeSettings.swift exactly: 4 neutral backgrounds, default
// is preset 100 ("White") — LIGHT canonical theme matching the wireframes.

data class BgPreset(
    val id: Int,
    val name: String,
    val bg: String,
    val surface: String,
    val card: String,
    val border: String,
    val text: String,
    val muted: String,
    val isLight: Boolean
)

class ThemeSettings(private val context: Context) : ViewModel() {

    companion object {
        // Accent presets — same order/colors as iOS ThemeSettings.accentPresets.
        val ACCENT_PRESETS = listOf(
            "#3B82F6", // Sky (default) — TRAQS Light
            "#7c3aed", // Purple
            "#10b981", // Green
            "#f59e0b", // Amber
            "#f43f5e", // Red
            "#FF1FB4", // Magenta
            "#06b6d4", // Cyan
            "#8b5cf6", // Violet
        )

        // Background presets — ONE light and ONE dark. Grey (10) and Black (12)
        // were retired: Grey sat close enough to White that the frosted surfaces
        // could not separate from it, and Black crushed the specular rim's
        // bottom lip into the page. The IDs are left unused rather than
        // renumbered so a saved server-side setting from another client still
        // resolves — anything not in this list falls back to White below.
        val BG_PRESETS = listOf(
            BgPreset(100, "White",    "#F4F6FA", "#FFFFFF", "#FFFFFF", "#E6E8EE", "#0B0B0C", "#6E6E73", true),
            BgPreset(11,  "Charcoal", "#1F1F1F", "#2A2A2A", "#333333", "#3F3F3F", "#E8E8E8", "#9CA3AF", false),
        )

        const val DEFAULT_BG_PRESET_ID = 100
        const val DEFAULT_ACCENT = "#3B82F6"

        private const val PREFS_NAME = "traqs_theme_prefs"
        private const val KEY_ACCENT = "accent"
        private const val KEY_BG_PRESET = "bg_preset"
        private const val KEY_FROSTED = "frosted_glass"
        private const val KEY_LIQUID = "liquid_background"
    }

    private val prefs = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)

    private val _accent = MutableStateFlow(prefs.getString(KEY_ACCENT, DEFAULT_ACCENT) ?: DEFAULT_ACCENT)
    val accent: StateFlow<String> = _accent.asStateFlow()

    // If a saved id isn't one of the 4 current presets, fall back to White.
    // Mirrors iOS migration of older preset ids.
    private val _bgPresetId = MutableStateFlow(
        prefs.getInt(KEY_BG_PRESET, DEFAULT_BG_PRESET_ID).let { saved ->
            if (BG_PRESETS.any { it.id == saved }) saved else DEFAULT_BG_PRESET_ID
        }
    )
    val bgPresetId: StateFlow<Int> = _bgPresetId.asStateFlow()

    val currentBgPreset: BgPreset get() = BG_PRESETS.firstOrNull { it.id == _bgPresetId.value } ?: BG_PRESETS[0]

    // Frosted Glass — mirrors the iOS Customize toggle. On, cards and panels are
    // the translucent glass the design system draws over the liquid background;
    // off, they flatten to opaque surfaces with a plain hairline. Buttons, the
    // nav pill and prompts stay glass either way, same as iOS. Default ON.
    private val _frostedGlass = MutableStateFlow(prefs.getBoolean(KEY_FROSTED, true))
    val frostedGlass: StateFlow<Boolean> = _frostedGlass.asStateFlow()

    // Liquid Motion — mirrors the iOS Customize toggle. On, the page canvas is
    // the drifting two-blob wash; off, it is the static ambient canvas (the same
    // ground with one soft glow in the upper right). Default ON.
    //
    // It is a real setting, not just taste: the wash runs two infinite
    // animations for as long as the app is open, and somebody on a long shift
    // may simply not want that.
    private val _liquidBackground = MutableStateFlow(prefs.getBoolean(KEY_LIQUID, true))
    val liquidBackground: StateFlow<Boolean> = _liquidBackground.asStateFlow()

    fun setLiquidBackground(on: Boolean) {
        _liquidBackground.value = on
        prefs.edit().putBoolean(KEY_LIQUID, on).apply()
    }

    fun setFrostedGlass(on: Boolean) {
        _frostedGlass.value = on
        prefs.edit().putBoolean(KEY_FROSTED, on).apply()
    }

    val isLightTheme: Boolean get() = currentBgPreset.isLight

    fun setAccent(hex: String) {
        _accent.value = hex
        prefs.edit().putString(KEY_ACCENT, hex).apply()
    }

    fun setBgPreset(id: Int) {
        _bgPresetId.value = id
        prefs.edit().putInt(KEY_BG_PRESET, id).apply()
    }

    fun reset() {
        setAccent(DEFAULT_ACCENT)
        setBgPreset(DEFAULT_BG_PRESET_ID)
        setFrostedGlass(true)
        setLiquidBackground(true)
    }
}
